app/scripts/controllers/network/fake-provider.ts

```diff
@@ -0,0 +1,224 @@
+  ) => {
+    return this.#handleSend(payload, callback);
+  };
+
+  sendAsync = <Params, Response>(
+    payload: JsonRpcRequest<Params>,
+    callback: (error: unknown, response?: JsonRpcResponse<Response>) => void,
+  ) => {
+    return this.#handleSend(payload, callback);
+  };
+
+  #handleSend<Params, Response>(
+    payload: JsonRpcRequest<Params>,
+    callback?: (error: unknown, response?: JsonRpcResponse<Response>) => void,
+  ) {
+    if (Array.isArray(payload)) {
+      throw new Error("Arrays aren't supported");
+    }
+
+    const index = this.#stubs.findIndex((stub) => {
+      return (
+        stub.request.method === payload.method &&
+        (!('params' in stub.request) ||
+          isDeepStrictEqual(stub.request.params, payload.params))
+      );
+    });
+
+    if (index === -1) {
+      const matchingCalledStubs = this.calledStubs.filter((stub) => {
+        return (
+          stub.request.method === payload.method &&
+          (!('params' in stub.request) ||
+            isDeepStrictEqual(stub.request.params, payload.params))
+        );
+      });
+      let message = `Could not find any stubs matching: ${inspect(payload, {
+        depth: null,
+      })}`;
+      if (matchingCalledStubs.length > 0) {
+        message += `\n\nIt appears the following stubs were defined, but have been called already:\n\n${inspect(
+          matchingCalledStubs,
+          { depth: null },
+        )}`;
+      }
+
+      throw new Error(message);
+    } else {
+      const stub = this.#stubs[index];
+
+      if (stub.discardAfterMatching !== false) {
+        this.#stubs.splice(index, 1);
+      }
+
+      if (callback === undefined) {
+        this.#handleRequest(stub, (error, response) => {
+          console.warn(
+            'got result or error but not doing anything with it',
+            error,
+            response,
+          );
+        });
+      } else if (stub.delay) {
+        originalSetTimeout(() => {
+          this.#handleRequest(stub, callback);
+        }, stub.delay);
+      } else {
+        this.#handleRequest(stub, callback);
+      }
+
+      this.calledStubs.push({ ...stub });
+    }
+  }
+
+  #handleRequest<Params, Response>(
+    stub: FakeProviderStub<Params, Response>,
+    callback: (error: unknown, response?: JsonRpcResponse<Response>) => void,
+  ) {
+    const promiseForBeforeCompleting =
+      stub.beforeCompleting === undefined
+        ? Promise.resolve()
+        : Promise.resolve(stub.beforeCompleting());
+
+    promiseForBeforeCompleting.then(() => {
+      if ('implementation' in stub) {
+        stub.implementation();
+      } else if ('response' in stub) {
+        if ('result' in stub.response) {
+          return callback(null, {
+            jsonrpc: '2.0',
+            id: 1,
+            result: stub.response.result,
+          });
+        } else if ('error' in stub.response) {
+          if (typeof stub.response.error === 'string') {
```

Comment:
Is this condition necessary? According to the stub type, this should always be a string

Reply:
Good point. Removed in [24c635b](https://github.com/MetaMask/metamask-extension/pull/18628/commits/24c635bdb4683c48912e9cc9dc9a9cbb8672c3bc).